### ISSUE TITLE
Crunchyroll

### ISSUE THREAD

**Issue by Commonman7** [FrostCo/AdvancedProfanityFilter#502]
Just went back onto chrunchyroll today and it is not filtering. I've tried it on chrome, firefox and edge browers with the same results. Thanks

**Comment by richardfrost** [comment]
Could you share a link to the video you tested with @Commonman7?

I don't personally have a subscription to Crunchy roll right now, so I've just been testing with this video: https://www.crunchyroll.com/watch/G7PU4VVV4/its-not-okay-the-spider-lures-sanji and it seems to be working okay on it for me. Does it work for you?

**Comment by Commonman7** [comment]
Haven’t tried your link yet but I will
Here’s what I was watching

https://crunchyroll.com/watch/GPWUKDJ19?utm_medium=ios&utm_source=share

**Comment by richardfrost** [comment]
Thanks for sending the link @Commonman7. When you get a chance, can you see if episode 1 works? https://www.crunchyroll.com/watch/GG1U2QV9E/ryusui 

If that does work for you, but episode 6 doesn't I'd guess they are doing something different between their free titles and the premium ones. If that's the case I'll re-sub for a month and take a look at it.

**Comment by Commonman7** [comment]
Just tried episode one and it didn’t work. I have also tried a different
device just in case and still no luck.

Thanks

**Comment by jaxmirox** [comment]
Hi Richard, I've been emailing back and forth.  I really appreciate you taking your time on this.  Thanks!

**Comment by richardfrost** [comment]
Thanks for all the help in tracking down the issue @Commonman7 and @jaxmirox. I have identified the issue and will be working through getting it fixed in an update. I'm aiming to finish it up this evening and will update when I have submitted it for review by the various stores.

**Comment by Commonman7** [comment]
Thanks for the update and for the great work you are doing.

**Comment by jaxmirox** [comment]
Thank you sir.  I really appreciate you doing all of this.

**Comment by richardfrost** [comment]
Thanks for all the encouragement and support @Commonman7 and @jaxmirox.

Sorry for the delay. This should be fixed in version [3.2.1](https://github.com/richardfrost/AdvancedProfanityFilter/releases/tag/3.2.1). Feel free to comment here if you have any feedback for this update.

It is available now for Firefox, with the other stores following as soon as it passes their review process.

**Comment by Commonman7** [comment]
Everything seems to be working fine.
Thanks again

**Comment by jinzjuunanagou** [comment]
I having problems getting this to work on crunchyroll, I see common got it to work. Just not sure how.

**Comment by richardfrost** [comment]
What are you seeing @jinzjuunanagou? First do you have the latest update (3.2.1)? Do you have any custom audio site config set for Crunchyroll (Options page > Audio tab)?

**Comment by jinzjuunanagou** [comment]
Ya I just downloaded it today so it should be latest version. But you might be right I be doing something wrong. So i have to custom audio for crunchy on the options page? I will give it a look.

**Comment by jinzjuunanagou** [comment]
Audio tab > Supported Sites Config > Copy/Pasted Crunchyroll still no luck.

**Comment by richardfrost** [comment]
@jinzjuunanagou, no, you shouldn't need custom audio config for it. You can remove it if you copied it. If you go to the "Help" tab you should be able to see the version in the top-right corner just to confirm you are on the latest (3.2.1). Also, which browser are you using?

Could you share a link to a video you are testing it on?

**Comment by jinzjuunanagou** [comment]
I am using Chrome.  Sure I am trying to get it to work on https://www.crunchyroll.com/watch/G14U4E83J/stone-world 1st episode on the 11:02 min mark the CC censors actually work but no video mute happens.

**Comment by Commonman7** [comment]
I logged onto Chrome tonight and it wouldn’t work. I’ve logged into Firefox
and it’s working fine.

**Comment by jaxmirox** [comment]
Working fine on chrome and firefox for me.

**Comment by jaxmirox** [comment]
<img width="1920" alt="dr stone english audio_english_cc" src="https://github.com/FrostCo/AdvancedProfanityFilter/assets/133458900/39e957f5-727c-48fd-848d-dca3ffe054f8">



Not sure if this helps

**Comment by richardfrost** [comment]
Thanks for all the reports and details. I'll be taking a look in a few minutes and I'll update when I have something to report.

**Comment by richardfrost** [comment]
It looks like you are right @jaxmirox, they appear to be using yet _another_ method for captions/subtitles for dubbed videos.

This custom config should work for the dubbed videos, but will prevent it working on the non-dubbed ones:

```json
{
  "www.crunchyroll.com": [
    {
      "iframe": true,
      "mode": "elementChild",
      "note": "Dubbed videos",
      "parentSelector": "div#vilosVttJs",
      "subtitleSelector": ":scope div > div",
      "tagName": "DIV"
    }
  ]
}
```

I'm going to see if I can get them to both work at the same time now.

**Comment by richardfrost** [comment]
Okay, I think this config should support all the variations (at least that I've come across):

```json
{
  "www.crunchyroll.com": [
    {
      "apfCaptions": true,
      "apfCaptionsSelector": "vilosVttJs",
      "displaySelector": "canvas#velocity-canvas",
      "externalSub": true,
      "externalSubTrackMode": "hidden",
      "externalSubVar": "window.v1config.media.playService.subtitles;;window.v1config.media.subtitles",
      "iframe": true,
      "mode": "cue",
      "videoCueLanguage": "en-US",
      "videoCueRequireShowing": false,
      "videoSelector": "video#player0"
    },
    {
      "iframe": true,
      "mode": "elementChild",
      "note": "Dubbed videos",
      "parentSelector": "div#vilosVttJs",
      "subtitleSelector": ":scope div > div",
      "tagName": "DIV"
    }
  ]
}
```

If anyone can give it a try and let me know how it works that would be great! Here are the videos I've been testing:

- [_free_] [**Japanese Audio**] - [One Piece - E1 - I’m Luffy! The Man Who’s Gonna Be King of the Pirates!](https://www.crunchyroll.com/watch/GR3VWXP96/im-luffy-the-man-whos-gonna-be-king-of-the-pirates) [0:08:52 (d***)]
- [_free_] [**Japanese Audio**] - [Dr. STONE - E1 - Stone World](https://www.crunchyroll.com/watch/G63K39W46/stone-world) [0:00:40 (h***)]
- [_free_] [**Dubbed English Audio**] - [Dr. STONE - E1 - Stone World](https://www.crunchyroll.com/watch/G14U4E83J/stone-world) (_Need to turn on `English [CC]`_) [0:00:40 (h***)]
- [_Premium_] [**Japanese Audio**] - [Dr. STONE - E6 - Treasure Box](https://www.crunchyroll.com/watch/GPWUKDJ19/treasure-box) [0:07:30 (d***)]

I included approximate timestamps for words that will be filtered with the default wordlist.

If this works, feel free to just use the above config until I can get another update published. I'll hold off with publishing this until I've heard back on how it works though.

**Comment by jinzjuunanagou** [comment]
I thank you much! This worked for me.

**Comment by richardfrost** [comment]
Thanks for reporting back @jinzjuunanagou, and once you get a chance to test it out @jaxmirox let me know and I'll work on getting this added (assuming it works for everyone).

**Comment by jaxmirox** [comment]
Will do.  I didn’t have a chance to last night but will try it today.  Thanks againSent from my iPhoneOn May 18, 2023, at 10:20 PM, Richard Frost ***@***.***> wrote:﻿
.Thanks for reporting back @jinzjuunanagou, and once you get a chance to test it out @jaxmirox let me know and I'll work on getting this added (assuming it works for everyone).

**Comment by jaxmirox** [comment]
This is working just fine for me as well.  Thanks again!

**Comment by richardfrost** [comment]
Thanks for reporting back @jaxmirox & @jinzjuunanagou. I'll get this added into the next release and let you know here in this issue when its been released.

**Comment by richardfrost** [comment]
This latest config has been included in version [3.2.2](https://github.com/richardfrost/AdvancedProfanityFilter/releases/tag/3.2.2). Feel free to comment here if you have any feedback for this update.

**Comment by richardfrost** [comment]
Crunchyroll support was updated in version [4.5.0](https://github.com/FrostCo/APF-Premium/releases/tag/4.5.0).

Crunchyroll made a lot of changes which now require you to change some of the options available on their player.

- For subbed videos: Choose to show `None` of the subtitles. The extension can't hide them anymore because they have made them part of the video stream itself. Even when set to `None`, you can use APF to show all the subtitles with filtering applied to them. If you don't do this, you will see 2 sets of subtitles, the one from Crunchyroll, and the one from the extension.
- For dubbed videos: Choose `[CC]` option when available, because several titles only include foreign text in the captions (meaning it may miss muting words you would like to be filtered).
